staging/src/k8s.io/sample-controller/controller.go

```diff
@@ -297,6 +308,7 @@ func (c *Controller) syncHandler(key string) error {
 	if foo.Spec.Replicas != nil && *foo.Spec.Replicas != *deployment.Spec.Replicas {
 		glog.V(4).Infof("Foo %s replicas: %d, deployment replicas: %d", name, *foo.Spec.Replicas, *deployment.Spec.Replicas)
 		deployment, err = c.kubeclientset.AppsV1().Deployments(foo.Namespace).Update(newDeployment(foo))
+		fooModified = true
```

Comment:
I believe this somewhat exposes a (two?) pre-existing bug in the sample controller. Right now we are not reconciling against deployment changes. So we are not accounting for the case where something else modified the deployment .Spec.Replicas. This could be a later change which modifies it. I think we should be listening for such changes. The other is a mutating webhook could alter the result of the create. We may want to at least add a section which explains to people that this could happen and helps them think through the problem. It would be great if you wanted to help fix that but at least can we file a bug for it?

Reply:
@cheftako Thanks for the comments. I have been thinking about your comments and have following questions/points:

- Why is reacting to deployment object changes separately handled from reacting to Foo resources? More generally, what would be a situation where a child object for a custom resource (such as deployment in case of Foo)  is updated independently/separately without updating/modifying the parent custom resource instance?

A related point is -- why is Foo Spec exposing the fact that there will be a Deployment object created when Foo is created? A realistic example would be where users only need to work with attributes that make sense at the level of 'Foo' and the foo controller takes care of creating the 'components' that are required for 'Foo'. Such components could be a deployment, a service, a configMap, etc. 

- I think if we look at the sample-controller primarily as controller of Foo resources that does not expose underlying implementation details in the Foo spec, then the points you mentioned above regarding deployment replicas getting updated independently of Foo resource becomes moot (also, component objects will usually not be updated separately from their parent). Then the focus of sample-controller remains only on managing Foo resources and reacting to only Foo changes (and not for any of its component objects like deployment).

- As far as webhooks that modify Foo resource, wouldn't their effect be similar to doing a manual update on an already created Foo resource? In both cases the controller will reconcile based on the updated Spec without doing any special treatment of webhooks. No?
So I did not follow why we have to explicitly call out the webhook point.


If above makes sense, then I think we need to modify sample-controller to remove 'deploymentName' from the Spec and rewrite the UpdateFuncs to only react to Foo changes. Note that we can leave 'replicas' attribute in the Spec. The reconciliation will be based on the replica count specified in the Spec and the current number of Pods created for that Foo deployment, or just the diff of 'availableReplicas' from the Status and the 'replicas' specified in the Spec.

What do you think?